pkg/genlib/generator_with_template.go

```diff
@@ -0,0 +1,529 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package genlib
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/rand"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Pallinder/go-randomdata"
+	"github.com/lithammer/shortuuid/v3"
+)
+
+var templateFieldMap map[string][]byte
+var trailingTemplate []byte
+
+// GeneratorWithTemplate is resolved at construction to a slice of emit functions
+type GeneratorWithTemplate struct {
+	emitFuncs []emitF
+}
+
+func generateTemplateFromField(cfg Config, fields Fields) []byte {
+	templateBuffer := bytes.NewBufferString("{")
+	for i, field := range fields {
+		fieldWrap := fieldValueWrapByType(field)
+		if fieldCfg, ok := cfg.GetField(field.Name); ok {
+			if fieldCfg.Value != nil {
+				fieldWrap = ""
+			}
+		}
+
+		fieldTrailer := []byte(",")
+		if i == len(fields)-1 {
+			fieldTrailer = []byte("}")
+		}
+
+		if strings.HasSuffix(field.Name, ".*") || field.Type == FieldTypeObject || field.Type == FieldTypeNested || field.Type == FieldTypeFlattened {
+			// This is a special case.  We are randomly generating keys on the fly
+			// Will set the json field name as "field.Name.N"
+			N := 5
+			for i := 0; i < N; i++ {
+				if string(fieldTrailer) == "}" && i < N-1 {
+					fieldTrailer = []byte(",")
+				}
+
+				fieldNameRoot := replacer.Replace(field.Name)
+				fieldTemplate := fmt.Sprintf(`"%s.%d": %s{{.%s.%d}}%s%s`, fieldNameRoot, i, fieldWrap, fieldNameRoot, i, fieldWrap, fieldTrailer)
+				templateBuffer.WriteString(fieldTemplate)
+			}
+		} else {
+			fieldTemplate := fmt.Sprintf(`"%s": %s{{.%s}}%s%s`, field.Name, fieldWrap, field.Name, fieldWrap, fieldTrailer)
+			templateBuffer.WriteString(fieldTemplate)
+		}
+	}
+
+	return templateBuffer.Bytes()
+}
+
+func NewGeneratorWithTemplate(template []byte, cfg Config, fields Fields) (*GeneratorWithTemplate, error) {
+	if len(template) == 0 {
+		template = generateTemplateFromField(cfg, fields)
```

Comment:
we can call the factory with an empty template: in this case the template will be generated as json from the `fields`, basically reproducing the same output of the "legacy" generator

this is useful in order to assess regression between legacy and template

when template generator will be performant enough we will remove the legacy and rely on this dynamic template generation